scripts/abins/instruments/instrument.py

```diff
@@ -19,14 +23,104 @@ class Instrument:
+
+        Setting parameters will be checked first and take priority over
+        higher-level instrument parameters.
+
+        e.g. if abins.parameters includes::
+
+            instruments = {
+                'THIS_INSTRUMENT': {'param': 'value-1',
+                                    'settings': {'SETTING':
+                                                 {'param': 'value-2'}}}
+            }
+
+        Then self.get_parameter('param') will return 'value-1' unless the
+        setting is 'SETTING', in which case it will return 'value-2'.
+
+        Args:
+            key: key to be searched in instrument parameters
+            default: Value to be returned if key is not found.
+        """
+        setting_parameters = self.get_setting_parameters()
+        if key in setting_parameters:
+            return setting_parameters[key]
+        else:
+            return self.get_parameters().get(key, default)
+
+    def get_parameters(self) -> Dict[str, Any]:
+        """Get parameters for this instrument from abins.parameters"""
+        return abins.parameters.instruments.get(self.get_name())
+
+    def get_setting(self) -> str:
         return self._setting
 
+    def get_setting_parameters(self) -> Dict[str, Any]:
+        """Get instrument parameters associated with the current setting
+
+        (These will be cached the first time this method is called. Instrument
+        settings are not mutable.)
+        """
+        if self._setting_parameters is None:
+            all_settings_data = self.get_parameters().get('settings')
+
+            if all_settings_data:
+                self._setting_parameters =  all_settings_data.get(
+                    self._check_setting(self.get_setting()))
+            else:
+                self._setting_parameters = {}
+
+        return self._setting_parameters
+
+    def get_min_wavenumber(self):
+        return self.get_parameter('min_wavenumber', default=abins.parameters.sampling['min_wavenumber'])
+
+    def get_max_wavenumber(self):
+        return self.get_parameter('max_wavenumber', default=abins.parameters.sampling['max_wavenumber'])
+
+    def get_energy_bin_width(self):
+        """Check abins.parameters and select appropriate energy bin in cm-1
+
+        (Decreasing) priority order is:
+
+        - abins.parameters.sampling['bin_width']: use this to override
+          instrument settings (e.g. for convergence test or rough-quality calculation)
+
+        - abins.parameters.instruments[INSTRUMENT]['energy_bin_width']
+
+        - abins.parameters.instruments[INSTRUMENT]['n_energy_bins']
+
+        - abins.parameters.sampling['default_n_energy_bins']
+
+        If using an n_energy_bins value, this is used to subdivide from zero to
+        max_wavenumber.
+
+        """
+        bin_width = abins.parameters.sampling['bin_width']
+        if not bin_width:
+            bin_width = self.get_parameter('energy_bin_width', default=None)
```

Comment:
will this and line 107 always return None if there is no value?

Reply:
`abins.parameters.sampling['bin_width']` is
- defined as None in the parameters file file, but
- actually a slightly special case; it is used by the Abins Algorithm to override other bin width settings. The default value for BinWidthInWavenumber (which will be passed through) is 1.0

The user is not allowed to use the Abins Algorithm to set bin_width to 0.0; validateInputs requires it to be between 1.0 and 10.0.

(I think this is overly restrictive, and also not an important option to offer through the Algorithm/GUI. I'd like to ultimately remove it and use the same scheme as Abins2D: Instruments have defaults, and if you want to set a custom value you need to mess around with abins.parameters. Maybe then we can provide some kind of sensible dictionary **kwargs option for advanced parameters to be passed around in a less hack-y way.)